cirq/circuits/expectation_value.py

```diff
@@ -0,0 +1,376 @@
+
+from typing import Dict, Union
+
+import numpy as np
+from cirq.google import XmonSimulator, XmonMeasurementGate
+from cirq.circuits.circuit import Circuit
+from cirq.ops import RotXGate, RotYGate, MeasurementGate, Pauli, PauliString, \
+    QubitId
+
+
+def measurement_gates(circuit: Circuit,
+                      q_dict: Dict[QubitId,
+                                   Union[int, str]]):
+    """
+    Helper Function to add measurement gates to qubits
+
+    Args:
+        circuit: cirq.Circuit to add measurements to
+        q_dict: dictionary of qubits, keys are integers or strings,
+            values are corresponding qubits
+
+    Returns:
+         circuit with measurements added
+    """
+    meas = [XmonMeasurementGate(key=index).on(
+        qubit) for qubit, index in q_dict.items()]
+
+    circuit.append(meas)
+
+    return circuit
+
+
+def expectation_from_sampling(circuit: Circuit,
+                              operator: Dict[PauliString, float],
+                              n_samples: int):
+    """
+    Calculates the expactation value of cost function (operator)
+
+    Args:
+        circuit: circ.Circuit type which prepares the state we are interested
+                 in calculating the expectation value of operator
+        operator: Operator input as a dictionary where keys are
+                    cirq.ops.PauliString type and values are coefficients
+                  example:
+                  paulistring = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.X, qubits[1]: (cirq.Pauli.Z)})
+                  paulistring2 = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.Z, qubits[2]: (cirq.Pauli.Y)})
+                  operator = {paulistring: 1.234, paulistring2: -5.678}
+        n_samples: number of runs for sampling
+
+    Returns:
+        expectation value of operator argument
+
+    """
+
+    qubits = list(circuit.all_qubits())
+    sim = XmonSimulator()
+
+    # remove measurements from circuit such that appropriate rotations
+    # can be applied, measurements are added later
+    circuit = no_meas(circuit)
+
+    # check if operator is quadratic in Pauli Z:
+    quadratic_z = True
+    zz = [Pauli.Z, Pauli.Z]
+    for key in operator.keys():
+        if (len(key) != 2):
+            quadratic_z = False
+            break
+        if list(key.values()) != zz:
+            quadratic_z = False
+            break
+
+    # if operator is only of the form ZZ then call helper function
+    if quadratic_z:
+
+        expectation = _expectation_from_sampling_assuming_quadratic(circuit,
+                                                                    operator,
+                                                                    n_samples)
+        return expectation
+
+
+    # For general operators, the expectation value by sampling
+    # over n_samples circuit runs is computed here
+    if not quadratic_z:
```

Comment:
This is a tautology due to the previous if-block. Drop this `if` line, just do its body unconditionally.